### PR TITLE
[PG12] Merge commit 'create complete graph function (#342)'

### PR DESCRIPTION
**Pull request by MuhammadTahaNaveed**
Merge [commit](https://github.com/apache/age/commit/dc5895ed0b6215ac29e15e66686de41af80a1538) into PG12.

Modifications with respect to age for PG12.

- heap_open -> table_open
- heap_close -> table_close
- change function return type to void from Oid wherever required

### CODE REVIEW

File: src/backend/commands/label_commands.c [hunk 1]
```diff
@@ -306,7 +306,8 @@ void create_label(char *graph_name, char *label_name, char label_type,
     // get a new "id" for the new label
     label_id = get_new_label_id(graph_oid, nsp_id);
 
-    insert_label(label_name, graph_oid, label_id, label_type, relation_id);
+    insert_label(label_name, graph_oid, label_id, label_type,
+                             relation_id, seq_name);
```
Comment (jrgemignani) [inline src/backend/commands/label_commands.c:310]: Please correct indentation

File: src/backend/utils/cache/ag_cache.c [hunk 2]
```diff
@@ -151,6 +171,12 @@ static label_cache_data *search_label_graph_oid_cache_miss(Oid graph,
 static void *label_graph_oid_cache_hash_search(uint32 graph, int32 id,
                                               HASHACTION action, bool *found);
 static label_cache_data *search_label_relation_cache_miss(Oid relation);
+static label_cache_data *search_label_seq_name_graph_cache_miss(Name name,
+                                                            Oid graph);
```
Comment (jrgemignani) [inline src/backend/utils/cache/ag_cache.c:175]: Please correct indentation

File: src/backend/utils/cache/ag_cache.c [hunk 3]
```diff
@@ -151,6 +171,12 @@ static label_cache_data *search_label_graph_oid_cache_miss(Oid graph,
...
+static void *label_seq_name_graph_cache_hash_search(Name name, Oid graph,
+                                                HASHACTION action,
```
Comment (jrgemignani) [inline src/backend/utils/cache/ag_cache.c:177]: Please correct indentation

File: src/backend/utils/cache/ag_cache.c [hunk 4]
```diff
@@ -151,6 +171,12 @@ static label_cache_data *search_label_graph_oid_cache_miss(Oid graph,
...
+                                                bool *found);
```
Comment (jrgemignani) [inline src/backend/utils/cache/ag_cache.c:178]: Please correct indentation

File: src/backend/utils/cache/ag_cache.c [hunk 5]
```diff
@@ -545,21 +579,42 @@ static void create_label_relation_cache(void)
...
+    MemSet(&hash_ctl, 0, sizeof(hash_ctl));
+    hash_ctl.keysize = sizeof(label_seq_name_graph_cache_key);
+    hash_ctl.entrysize = sizeof(label_seq_name_graph_cache_entry);
+
+    /*
+     * Please see the comment of hash_create() for the nelem value 16 here.
+     * HASH_BLOBS flag is set because the key for this hash is fixed-size.
+     */
+    label_seq_name_graph_cache_hash = hash_create("ag_label (seq_name, graph) cache",
+                                              16, &hash_ctl,
```
Comment (jrgemignani) [inline src/backend/utils/cache/ag_cache.c:595]: Please correct indentation

File: src/backend/utils/cache/ag_cache.c [hunk 6]
```diff
@@ -545,21 +579,42 @@ static void create_label_relation_cache(void)
...
+                                              HASH_ELEM | HASH_BLOBS);
```
Comment (jrgemignani) [inline src/backend/utils/cache/ag_cache.c:596]: Please correct indentation

File: src/backend/utils/cache/ag_cache.c [hunk 7]
```diff
@@ -929,6 +1039,88 @@ static label_cache_data *search_label_relation_cache_miss(Oid relation)
...
+
+    AssertArg(name);
+    AssertArg(OidIsValid(graph));
+
+    initialize_caches();
+
+    namestrcpy(&name_key, name);
+    entry = label_seq_name_graph_cache_hash_search(&name_key, graph, HASH_FIND,
+                                               NULL);
+    if (entry)
```
Comment (jrgemignani) [inline src/backend/utils/cache/ag_cache.c:1055]: All if/else use {} even for just one command.

File: src/backend/utils/cache/ag_cache.c [hunk 8]
```diff
@@ -929,6 +1039,88 @@ static label_cache_data *search_label_relation_cache_miss(Oid relation)
...
+    {
+        systable_endscan(scan_desc);
+        table_close(ag_label, AccessShareLock);
+
+        return NULL;
+    }
+
+    // get a new entry
+    entry = label_seq_name_graph_cache_hash_search(name, graph, HASH_ENTER,
+                                               &found);
```
Comment (jrgemignani) [inline src/backend/utils/cache/ag_cache.c:1100]: Please correct indentation

File: src/backend/utils/graph_generation.c [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+            eid = nextval_internal(edge_seq_id, true);
+            object_graph_id = make_graphid(edge_label_id, eid);
+
+            start_vertex_graph_id = make_graphid(vtx_label_id, start_vid);
+            end_vertex_graph_id = make_graphid(vtx_label_id, end_vid);
+
+            props = create_empty_agtype();
+
+            insert_edge_simple(graph_id, edge_name_str,
+                            object_graph_id, start_vertex_graph_id,
```
Comment (jrgemignani) [inline src/backend/utils/graph_generation.c:190]: Please correct indentation

File: src/backend/utils/graph_generation.c [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+                            end_vertex_graph_id, props);
```
Comment (jrgemignani) [inline src/backend/utils/graph_generation.c:191]: Please correct indentation

**Review by jrgemignani — CHANGES_REQUESTED** [review]
I realize that some of these changes are due to the original committer.  Still, although minor, we should fix them.

**Comment by jrgemignani** [comment]
@MuhammadTahaNaveed Other than the minor cosmetic adjustments above, the PR builds locally and looks good.

**Comment by MuhammadTahaNaveed** [comment]
> @MuhammadTahaNaveed Other than the minor cosmetic adjustments above, the PR builds locally and looks good.

I have fixed them.

**Comment by jrgemignani** [comment]
@MuhammadTahaNaveed Good job! Go ahead and squash your commit into the original commit and I'll merge it in.

**Comment by MuhammadTahaNaveed** [comment]
> @MuhammadTahaNaveed Good job! Go ahead and squash your commit into the original commit and I'll merge it in.
Thanks! I have squashed it.

**Pull request merged**